plugin/storage/es/esRollover.py

```diff
@@ -118,9 +119,12 @@ def create_aliases(client, alias_name, archive_index_name):
     """
     ilo = curator.IndexList(client)
     ilo.filter_by_regex(kind='regex', value='^'+archive_index_name+'$')
-    alias = curator.Alias(client=client, name=alias_name)
     for index in ilo.working_list():
         print("Adding index {} to alias {}".format(index, alias_name))
+    if re.search(r'write', alias_name):
```

Comment:
To address the failing tests due to unsupported `is_write_index` setting, we can try adding `and esVersion == 7` and pass in the `esVersion` as a param to `create_aliases`.

Reply:
Makes sense. Will do

Reply:
While going through the failed test - https://github.com/jaegertracing/jaeger/blob/master/plugin/storage/es/factory_test.go#L166, I realized that  mappings/jaeger-service-7.json and mappings/jaeger-span-7.json are being used by functions in factory.go as well. 

https://github.com/jaegertracing/jaeger/blob/master/plugin/storage/es/factory.go#L190

I have added Prefix and aliases in the settings of the template, which are expected to passed while parsing the template.
https://github.com/jaegertracing/jaeger/blob/24fb6bc0f05d865eb113c28f961ced6bf62e7793/plugin/storage/es/factory.go#L212

I hope this wont impact the functionality of ES plugin itself. @albertteoh kindly share your thoughts

Reply:
Still going through the code, previously I have looked at esRollover.py in an isolated manner. Dint go through rest of the code. 
After initial analysis of the ES plugin code, I  believe we can use the index templates created by ES plugin by default and go away with esRollover script ( for ILM implementation, all it does is to create over riding index templates). Let me make some changes - would need some time.

Reply:
Yup, good point. If modifying the existing default index template created by the ES plugin does not impact existing behaviour with manual init+rollover of indices, but also satisfies the requirement for enabling ILM, I think that would be a cleaner approach. After all, it seems that's what ES is currently doing anyway; it's merging the index template created by the ES plugin with the ilm template, and from what I've seen so far, both ILM and the manual rollover scripts are working as expected.

Reply:
@albertteoh  - Have DMed on gitter with few observations. Please let me know your thoughts. I will post the approach here - once we have clarity

Reply:
As discussed, I would be working on below things -

- Have a flag to enable ILM in index template ( We decided to use same index template, rather than multiple copies jus for ILM)
- Use Go templates for templating index template ( As of now, we replace for strings in the  files. Link to code below )

https://github.com/jaegertracing/jaeger/blob/5c3b53bd780c7dc23392fceb08c68f60c13997af/plugin/storage/es/factory.go#L212
- Add integration tests for enabling ILM.

@albertteoh  - Kindly add if I missed something

Reply:
Also, as discusssed, given the same string replacement logic is needed for both Go and Python i.e.: 

https://github.com/jaegertracing/jaeger/blob/58d3f5ccbf95c59fe329b79a2729fe9d090655e9/plugin/storage/es/esRollover.py#L177

A similar templating solution for Python would be nice (e.g. Jinja).

Reply:
@albertteoh  - Implemented below tasks -
>- Have a flag to enable ILM in index template ( We decided to use same index template, rather than multiple copies jus for ILM)
>- Use templates for templating index template ( As of now, we replace for strings in the files. Link to code below ) 
Used django style templates as we needed a common templates which could be rendered in python as well as golang.  ref: github.com/flosch/pongo2

It would still support older way of cleaning and rollover. TestIndexCleaner integration tests are failing, not sure if they are related to my changes - would spend time on those.
 
Would work on adding integration test for ILM.

 Needed some guidance on below points - 

- Do we need a check that use_ilm param is always used with use_aliases. If so where to add that check?
- As of now I am not handling/throwing the error while rendering the template -

https://github.com/bhiravabhatla/jaeger/blob/fe5938fb480f4ce896861680188c898deadd7233/plugin/storage/es/factory.go#L212
Can I add a panic here?

- Do we need to have a check if ILM policy is created in es_rollover script

Reply:
Thanks for the update, @bhiravabhatla.

> Can I add a panic here?

I think we should return the error instead of panicking, and let it bubble up the call stack.

> Do we need to have a check if ILM policy is created in es_rollover script

If you mean to prevent the manual rollover script from stepping on the toes ILM's automated rollover (or vice versa), then yes, I think we should check if ILM policy is created to avoid any unexpected behaviour or states. In other words, I think we should use either one approach (manual rollover & clean scripts) or the other (ES's built-in ILM feature), but not both.

Reply:
@albertteoh  
> If you mean to prevent the manual rollover script from stepping on the toes ILM's automated rollover (or vice versa), then yes, I think we should check if ILM policy is created to avoid any unexpected behaviour or states. In other words, I think we should use either one approach (manual rollover & clean scripts) or the other (ES's built-in ILM feature), but not both.
Sorry, the question was not clear enough. I dont think we can stop users from manual rollover anyways - as one with access to ES can always call rollover API. Also, ILM is intelligent enough to skip rolling over indices which already rolled over manually. Ref - https://github.com/elastic/elasticsearch/pull/47324

The intention behind my question was, as we have a hardcoded ILM name in index template (jaeger-ilm-policy), before jaeger starts up, this ILM policy must be present in ES. If its not present the indices would show ILM errors - although this wont impact Jaeger in any ways.

I see 2 ways to address this - 
* This step could be documented as prerequisite and  we can have a check in esRollover.py - before creating templates - whether we have the ILM already created. If not exit with error saying please create ILM in ES.
* We create a bare minimum ILM ourselves in esRollover. and Users can modify it later

Reply:
I see, thanks for clarifying @bhiravabhatla. I prefer this option:
>This step could be documented as prerequisite and we can have a check in esRollover.py - before creating templates - whether we have the ILM already created. If not exit with error saying please create ILM in ES.

... mainly because if we provide a bare minimum, it could silently rollover and delete indices (or never delete) without the users' knowledge, until it's too late.

I think it would be good to do the same in the ES storage plugin within Jaeger collector as well. That is, if `ES_USE_ILM == true`:
- Is the jaeger ILM created?
- Can & should we validate that the `init` script has been called? IIUC, if `init` isn't called, ILM rollovers would fail.

Reply:
Making ILM name configurable makes sense - also validating if the provided ILM name exists. I ll make these changes.

As far as having the flag in collector, i have shared my thoughts on below thread. https://github.com/jaegertracing/jaeger/pull/2454#discussion_r538345684

Reply:
For first iteration/initial draft - I am thinking of having ILM policy name hardcoded to jaeger-ilm-policy. Will add a check if this ILM exists in python script (esRollover). 

I am also adding a bool flag to collector ES_UseILM == true as discussed in below threads.

Reply:
Implemented above feedback in latest commit. Kindly go through @albertteoh